core/example/longrange/ewald.cpp

```diff
@@ -103,210 +76,428 @@ double TEwald::compute( ParticleList &particles, double lx, double ly,
+        f( idx, 0 ) = 0.0;
+        f( idx, 1 ) = 0.0;
+        f( idx, 2 ) = 0.0;
+    };
     Kokkos::parallel_for( Kokkos::RangePolicy<ExecutionSpace>( 0, n_max ),
                           init_p );
     Kokkos::fence();
 
     double alpha = _alpha;
-    double k_max = _k_max;
     double r_max = _r_max;
     double eps_r = _eps_r;
 
-#ifdef Cabana_ENABLE_Cuda
-    Kokkos::View<int *, MemorySpace> k_max_int( "k_max_int", 3 );
-    for ( auto i = 0; i < 3; ++i )
-    {
-        k_max_int[i] = _k_max_int[i];
-    } // TODO: enhancement - use Views instead. No need for macros.
-#else
-    int *k_max_int = &( _k_max_int[0] );
-#endif
+    auto start_time_kf = std::chrono::high_resolution_clock::now();
+    EwaldUkFunctor ukf( particles, _k_max, _alpha, lx, ly, lz );
+    std::cout << "k-space potentials functor created" << std::endl;
+    int k_int = std::ceil( _k_max );
+    int n_k = 8 * k_int * k_int * k_int + 12 * k_int * k_int + 6 * k_int + 1;
+    Kokkos::parallel_reduce( n_k, ukf, Uk );
+    std::cout << "k-space potentials computed" << std::endl;
+    Kokkos::fence();
+    auto end_time_kf = std::chrono::high_resolution_clock::now();
+    auto start_time_kff = std::chrono::high_resolution_clock::now();
+    EwaldUkForcesFunctor<ExecutionSpace> uk_fi( r, q, f, _k_max, _alpha, n_k,
+                                                lx, ly, lz );
+    std::cout << "k-space forces functor created" << std::endl;
+    Kokkos::parallel_for(
+        Kokkos::TeamPolicy<ExecutionSpace>( n_max, Kokkos::AUTO ), uk_fi );
+    Kokkos::fence();
+
+    auto end_time_kff = std::chrono::high_resolution_clock::now();
+    auto elapsed_time_kf = end_time_kf - start_time_kf;
+    auto elapsed_time_kff = end_time_kff - start_time_kff;
+    auto ns_elapsed_kf =
+        std::chrono::duration_cast<std::chrono::nanoseconds>( elapsed_time_kf );
+    auto ns_elapsed_kff = std::chrono::duration_cast<std::chrono::nanoseconds>(
+        elapsed_time_kff );
+
+    std::cout << "k-space contribution: "
+              << ( ( ns_elapsed_kf + ns_elapsed_kff ).count() / 1000000000.0 )
+              << " s = "
+              << " potential: " << ( ns_elapsed_kf.count() / 1000000000.0 )
+              << " s + "
+              << " forces: " << ( ns_elapsed_kff.count() / 1000000000.0 )
+              << " s " << Uk << " " << std::endl;
+
     // computation real-space contribution
-    Kokkos::parallel_reduce(
-        Kokkos::RangePolicy<ExecutionSpace>( 0, n_max ),
-        KOKKOS_LAMBDA( int idx, double &Ur_part ) {
-            double d[SPACE_DIM];
-            double k;
-            // For each particle with charge q, the real space contribution to
-            // energy is Ur_part = 0.5*q*SUM_i(q_i*erfc(alpha*dist)/dist) The
-            // sum is over all other particles in the cell and in neighboring
-            // images up to some real-space cutoff distance "r_max"
-            //
-            // Self-energy terms are included then corrected for later, with one
```

Comment:
I'd like to keep some version of these detailed comments. Could some description be added in the new LinkedCellList implementation?

Reply:
Wrote a short bit about it, at the beginning of the branch